rflx/typing_.py

```diff
@@ -0,0 +1,254 @@
+import typing as ty
+from abc import abstractmethod
+from dataclasses import dataclass
+
+
+@dataclass(frozen=True)
+class Bounds:
+    lower: ty.Optional[int]
+    upper: ty.Optional[int]
+
+    def __post_init__(self) -> None:
+        assert self.lower is None or self.upper is None or self.lower <= self.upper
+
+    def __bool__(self) -> bool:
+        return self.lower is not None and self.upper is not None
+
+    def __contains__(self, item: object) -> bool:
+        if isinstance(item, int):
+            if self.lower is None or self.upper is None:
+                return False
+            return self.lower <= item <= self.upper
+        if isinstance(item, Bounds):
+            if not self and not item:
+                return True
+            if (
+                self.lower is not None
+                and self.upper is not None
+                and item.lower is not None
+                and item.upper is not None
+            ):
+                return self.lower <= item.lower and self.upper >= item.upper
+        return False
+
+    def __str__(self) -> str:
+        if not self:
+            return "undefined"
+        if self.lower == self.upper:
+            return str(self.lower)
+        return f"{self.lower} .. {self.upper}"
+
+    @staticmethod
+    def union(left: "Bounds", right: "Bounds") -> "Bounds":
+        if left.lower is None or left.upper is None or right.lower is None or right.upper is None:
+            return Bounds(None, None)
+        return Bounds(min(left.lower, right.lower), max(left.upper, right.upper))
```

Comment:
Yes, this is intended. Bounds are only used for integers, and integers cannot have any gaps in their value range. The use case for this function is to determine the bounds of an aggregate of integers, so that it can be checked if the aggregate is compatible to a concrete array type (i.e. all elements of the aggregate are in the bounds of the element type of the array).

Reply:
I see.